Native build failed due to UnresolvedElementException: Discovered unresolved method during parsing: io.quarkus.hibernate.orm.panache.PanacheEntityBase.count()

This is with Quarkus 1.1.

```
[INFO] --- quarkus-maven-plugin:1.1.0.Final:build (default) @ quarkus-qute ---
[WARNING] [io.quarkus.creator.curator.Curator] These dependencies are not recommended:
	- jakarta.xml.bind:jakarta.xml.bind-api should be replaced by org.jboss.spec.javax.xml.bind:jboss-jaxb-api_2.3_spec
You might end up with two different versions of the same classes or with an artifact you shouldn't have in your classpath.
[INFO] [org.jboss.threads] JBoss Threads version 3.0.0.Final
[INFO] [org.hibernate.jpa.boot.internal.PersistenceXmlParser] HHH000318: Could not find any META-INF/persistence.xml file in the classpath
[INFO] [org.hibernate.Version] HHH000412: Hibernate Core {5.4.10.Final}
[INFO] [io.quarkus.deployment.pkg.steps.JarResultBuildStep] Building native image source jar: /Users/gunnar/Development/quarkus/quarkus-qute/target/quarkus-qute-1.0.0-SNAPSHOT-native-image-source-jar/quarkus-qute-1.0.0-SNAPSHOT-runner.jar
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] Building native image from /Users/gunnar/Development/quarkus/quarkus-qute/target/quarkus-qute-1.0.0-SNAPSHOT-native-image-source-jar/quarkus-qute-1.0.0-SNAPSHOT-runner.jar
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] Running Quarkus native-image plugin on GraalVM Version 19.2.1 CE
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] docker run -v /Users/gunnar/Development/quarkus/quarkus-qute/target/quarkus-qute-1.0.0-SNAPSHOT-native-image-source-jar:/project:z --rm quay.io/quarkus/ubi-quarkus-native-image:19.2.1 -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-DCoordinatorEnvironmentBean.transactionStatusManagerEnable=false -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 -H:ResourceConfigurationFiles=resources-config.json --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy$BySpaceAndTime -jar quarkus-qute-1.0.0-SNAPSHOT-runner.jar -J-Djava.util.concurrent.ForkJoinPool.common.parallelism=1 -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -H:+AddAllCharsets -H:EnableURLProtocols=http,https --enable-all-security-services -H:+JNI --no-server -H:-UseServiceLoaderFeature -H:+StackTrace quarkus-qute-1.0.0-SNAPSHOT-runner
[quarkus-qute-1.0.0-SNAPSHOT-runner:19]    classlist:  45,459.72 ms
[quarkus-qute-1.0.0-SNAPSHOT-runner:19]        (cap):   2,092.14 ms
[quarkus-qute-1.0.0-SNAPSHOT-runner:19]        setup:   5,360.36 ms
17:28:11,883 INFO  [org.hib.Version] HHH000412: Hibernate Core {5.4.10.Final}
17:28:11,998 INFO  [org.hib.ann.com.Version] HCANN000001: Hibernate Commons Annotations {5.1.0.Final}
17:28:12,103 INFO  [org.hib.dia.Dialect] HHH000400: Using dialect: io.quarkus.hibernate.orm.runtime.dialect.QuarkusPostgreSQL95Dialect
17:28:14,836 INFO  [org.jbo.threads] JBoss Threads version 3.0.0.Final
[quarkus-qute-1.0.0-SNAPSHOT-runner:19]     analysis: 137,334.02 ms
Error: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: io.quarkus.hibernate.orm.panache.PanacheEntityBase.count(). To diagnose the issue you can use the --allow-incomplete-classpath option. The missing method is then reported at run time when it is accessed the first time.
Detailed message:
Trace:
	at parsing io.quarkus.hibernate.orm.panache.PanacheEntityBase_ValueResolver.resolve(PanacheEntityBase_ValueResolver.zig:109)
Call path from entry point to io.quarkus.hibernate.orm.panache.PanacheEntityBase_ValueResolver.resolve(EvalContext):
	at io.quarkus.hibernate.orm.panache.PanacheEntityBase_ValueResolver.resolve(PanacheEntityBase_ValueResolver.zig:68)
	at io.quarkus.qute.EvaluatorImpl.resolve(EvaluatorImpl.java:92)
	at io.quarkus.qute.EvaluatorImpl.lambda$resolve$2(EvaluatorImpl.java:94)
	at io.quarkus.qute.EvaluatorImpl$$Lambda$1141/525500644.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:966)
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:940)
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:456)
	at com.oracle.svm.core.jdk.RuntimeSupport.executeHooks(RuntimeSupport.java:144)
	at com.oracle.svm.core.jdk.RuntimeSupport.executeStartupHooks(RuntimeSupport.java:89)
	at com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:143)
	at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:186)
	at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)

com.oracle.svm.core.util.UserError$UserException: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: io.quarkus.hibernate.orm.panache.PanacheEntityBase.count(). To diagnose the issue you can use the --allow-incomplete-classpath option. The missing method is then reported at run time when it is accessed the first time.
Detailed message:
Trace:
	at parsing io.quarkus.hibernate.orm.panache.PanacheEntityBase_ValueResolver.resolve(PanacheEntityBase_ValueResolver.zig:109)
Call path from entry point to io.quarkus.hibernate.orm.panache.PanacheEntityBase_ValueResolver.resolve(EvalContext):
	at io.quarkus.hibernate.orm.panache.PanacheEntityBase_ValueResolver.resolve(PanacheEntityBase_ValueResolver.zig:68)
	at io.quarkus.qute.EvaluatorImpl.resolve(EvaluatorImpl.java:92)
	at io.quarkus.qute.EvaluatorImpl.lambda$resolve$2(EvaluatorImpl.java:94)
	at io.quarkus.qute.EvaluatorImpl$$Lambda$1141/525500644.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:966)
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:940)
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:456)
	at com.oracle.svm.core.jdk.RuntimeSupport.executeHooks(RuntimeSupport.java:144)
	at com.oracle.svm.core.jdk.RuntimeSupport.executeStartupHooks(RuntimeSupport.java:89)
	at com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:143)
	at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:186)
	at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)

	at com.oracle.svm.core.util.UserError.abort(UserError.java:75)
	at com.oracle.svm.hosted.FallbackFeature.reportAsFallback(FallbackFeature.java:223)
	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:737)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:526)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:444)
	at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: io.quarkus.hibernate.orm.panache.PanacheEntityBase.count(). To diagnose the issue you can use the --allow-incomplete-classpath option. The missing method is then reported at run time when it is accessed the first time.
Detailed message:
Trace:
	at parsing io.quarkus.hibernate.orm.panache.PanacheEntityBase_ValueResolver.resolve(PanacheEntityBase_ValueResolver.zig:109)
Call path from entry point to io.quarkus.hibernate.orm.panache.PanacheEntityBase_ValueResolver.resolve(EvalContext):
	at io.quarkus.hibernate.orm.panache.PanacheEntityBase_ValueResolver.resolve(PanacheEntityBase_ValueResolver.zig:68)
	at io.quarkus.qute.EvaluatorImpl.resolve(EvaluatorImpl.java:92)
	at io.quarkus.qute.EvaluatorImpl.lambda$resolve$2(EvaluatorImpl.java:94)
	at io.quarkus.qute.EvaluatorImpl$$Lambda$1141/525500644.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:966)
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:940)
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:456)
	at com.oracle.svm.core.jdk.RuntimeSupport.executeHooks(RuntimeSupport.java:144)
	at com.oracle.svm.core.jdk.RuntimeSupport.executeStartupHooks(RuntimeSupport.java:89)
	at com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:143)
	at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:186)
	at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)

	at com.oracle.graal.pointsto.constraints.UnsupportedFeatures.report(UnsupportedFeatures.java:130)
	at com.oracle.graal.pointsto.BigBang.finish(BigBang.java:565)
	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:688)
	... 7 more
Caused by: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: io.quarkus.hibernate.orm.panache.PanacheEntityBase.count(). To diagnose the issue you can use the --allow-incomplete-classpath option. The missing method is then reported at run time when it is accessed the first time.
	at com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.reportUnresolvedElement(SharedGraphBuilderPhase.java:238)
	at com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.handleUnresolvedMethod(SharedGraphBuilderPhase.java:228)
	at com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.handleUnresolvedInvoke(SharedGraphBuilderPhase.java:182)
	at org.graalvm.compiler.java.BytecodeParser.genInvokeVirtual(BytecodeParser.java:1612)
	at org.graalvm.compiler.java.BytecodeParser.processBytecode(BytecodeParser.java:5123)
	at org.graalvm.compiler.java.BytecodeParser.iterateBytecodesForBlock(BytecodeParser.java:3267)
	at org.graalvm.compiler.java.BytecodeParser.processBlock(BytecodeParser.java:3074)
	at org.graalvm.compiler.java.BytecodeParser.build(BytecodeParser.java:976)
	at org.graalvm.compiler.java.BytecodeParser.buildRootMethod(BytecodeParser.java:870)
	at org.graalvm.compiler.java.GraphBuilderPhase$Instance.run(GraphBuilderPhase.java:84)
	at org.graalvm.compiler.phases.Phase.run(Phase.java:49)
	at org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:197)
	at org.graalvm.compiler.phases.Phase.apply(Phase.java:42)
	at org.graalvm.compiler.phases.Phase.apply(Phase.java:38)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.parse(MethodTypeFlowBuilder.java:221)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:340)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.doParse(MethodTypeFlow.java:310)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureParsed(MethodTypeFlow.java:300)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.addContext(MethodTypeFlow.java:107)
	at com.oracle.graal.pointsto.DefaultAnalysisPolicy$DefaultVirtualInvokeTypeFlow.onObservedUpdate(DefaultAnalysisPolicy.java:191)
	at com.oracle.graal.pointsto.flow.TypeFlow.notifyObservers(TypeFlow.java:343)
	at com.oracle.graal.pointsto.flow.TypeFlow.update(TypeFlow.java:385)
	at com.oracle.graal.pointsto.BigBang$2.run(BigBang.java:510)
	at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$execute$0(CompletionExecutor.java:171)
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
	... 4 more
Error: Image build request failed with exit status 1
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  05:27 min
[INFO] Finished at: 2020-01-03T18:30:30+01:00
[INFO] ------------------------------------------------------------------------
```

--- a/independent-projects/qute/core/src/main/java/io/quarkus/qute/TemplateData.java
+++ b/independent-projects/qute/core/src/main/java/io/quarkus/qute/TemplateData.java
@@ -32,6 +32,11 @@ public @interface TemplateData {
     String[] ignore() default {};
 
     /**
+     * If set to true do not automatically analyze subclasses.
+     */
+    boolean ignoreSubclasses() default false;
+
+    /**
      * If set to true include only properties: instance fields and methods without params.
      */
     boolean properties() default false;

--- a/independent-projects/qute/generator/src/main/java/io/quarkus/qute/generator/ValueResolverGenerator.java
+++ b/independent-projects/qute/generator/src/main/java/io/quarkus/qute/generator/ValueResolverGenerator.java
@@ -70,6 +70,10 @@ public class ValueResolverGenerator {
     private static final String GET_PREFIX = "get";
     private static final String IS_PREFIX = "is";
 
+    private static final String IGNORE_SUBCLASSES = "ignoreSubclasses";
+    private static final String IGNORE = "ignore";
+    private static final String PROPERTIES = "properties";
+
     private final Set<String> analyzedTypes;
     private final Set<String> generatedTypes;
     private final IndexView index;
@@ -99,12 +103,18 @@ public class ValueResolverGenerator {
             return;
         }
         analyzedTypes.add(clazzName);
+        boolean ignoreSubclasses = false;
 
         // @TemplateData declared on class has precedence
         AnnotationInstance templateData = clazz.classAnnotation(TEMPLATE_DATA);
         if (templateData == null) {
             // Try to find @TemplateData declared on other classes
             templateData = uncontrolled.get(clazz);
+        } else {
+            AnnotationValue ignoreSubclassesValue = templateData.value(IGNORE_SUBCLASSES);
+            if (ignoreSubclassesValue != null) {
+                ignoreSubclasses = ignoreSubclassesValue.asBoolean();
+            }
         }
 
         Predicate<AnnotationTarget> filters = initFilters(templateData);
@@ -130,7 +140,7 @@ public class ValueResolverGenerator {
 
         valueResolver.close();
 
-        if (!clazz.superName().equals(OBJECT)) {
+        if (!ignoreSubclasses && !clazz.superName().equals(OBJECT)) {
             ClassInfo superClass = index.getClassByName(clazz.superClassType().name());
             if (superClass != null) {
                 generate(superClass);
@@ -577,7 +587,7 @@ public class ValueResolverGenerator {
         Predicate<AnnotationTarget> filter = ValueResolverGenerator::defaultFilter;
         if (templateData != null) {
             // @TemplateData is present
-            AnnotationValue ignoreValue = templateData.value("ignore");
+            AnnotationValue ignoreValue = templateData.value(IGNORE);
             if (ignoreValue != null) {
                 List<Pattern> ignore = Arrays.asList(ignoreValue.asStringArray()).stream().map(Pattern::compile)
                         .collect(Collectors.toList());
@@ -589,7 +599,7 @@ public class ValueResolverGenerator {
                     }
                 });
             }
-            AnnotationValue propertiesValue = templateData.value("properties");
+            AnnotationValue propertiesValue = templateData.value(PROPERTIES);
             if (propertiesValue != null && propertiesValue.asBoolean()) {
                 filter = filter.and(ValueResolverGenerator::propertiesFilter);
             }
@@ -612,16 +622,12 @@ public class ValueResolverGenerator {
         switch (target.kind()) {
             case METHOD:
                 MethodInfo method = target.asMethod();
-                if (method.name().equals("<init>")
-                        || method.name().equals("<clinit>") || isSynthetic(method.flags())
-                        || !Modifier.isPublic(method.flags())
-                        || method.returnType().kind() == org.jboss.jandex.Type.Kind.VOID) {
-                    return false;
-                } else {
-                    return true;
-                }
+                return Modifier.isPublic(method.flags()) && !Modifier.isStatic(method.flags()) && !isSynthetic(method.flags())
+                        && method.returnType().kind() != org.jboss.jandex.Type.Kind.VOID && !method.name().equals("<init>")
+                        && !method.name().equals("<clinit>");
             case FIELD:
-                return Modifier.isPublic(target.asField().flags()) && !Modifier.isStatic(target.asField().flags());
+                FieldInfo field = target.asField();
+                return Modifier.isPublic(field.flags()) && !Modifier.isStatic(field.flags());
             default:
                 throw new IllegalArgumentException("Unsupported annotation target");
         }

--- a/independent-projects/qute/generator/src/test/java/io/quarkus/qute/generator/BaseService.java
+++ b/independent-projects/qute/generator/src/test/java/io/quarkus/qute/generator/BaseService.java
@@ -1,0 +1,9 @@
+package io.quarkus.qute.generator;
+
+public class BaseService {
+
+    public String getBase() {
+        return "oops";
+    }
+
+}

--- a/independent-projects/qute/generator/src/test/java/io/quarkus/qute/generator/PublicMyService.java
+++ b/independent-projects/qute/generator/src/test/java/io/quarkus/qute/generator/PublicMyService.java
@@ -2,8 +2,8 @@ package io.quarkus.qute.generator;
 
 import io.quarkus.qute.TemplateData;
 
-@TemplateData
-public class PublicMyService {
+@TemplateData(ignoreSubclasses = true)
+public class PublicMyService extends BaseService {
 
     public String getName() {
         return "Martin";
@@ -11,6 +11,10 @@ public class PublicMyService {
 
     String getSurname() {
         return "Kouba";
+    }
+
+    public static boolean isStatic() {
+        return true;
     }
 
 }

--- a/independent-projects/qute/generator/src/test/java/io/quarkus/qute/generator/SimpleGeneratorTest.java
+++ b/independent-projects/qute/generator/src/test/java/io/quarkus/qute/generator/SimpleGeneratorTest.java
@@ -1,14 +1,13 @@
 package io.quarkus.qute.generator;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
 
 import io.quarkus.qute.Engine;
 import io.quarkus.qute.EvalContext;
 import io.quarkus.qute.Expression;
-import io.quarkus.qute.IfSectionHelper;
 import io.quarkus.qute.ImmutableList;
 import io.quarkus.qute.ValueResolver;
-import io.quarkus.qute.ValueResolvers;
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.Collections;
@@ -28,7 +27,8 @@ public class SimpleGeneratorTest {
     @BeforeAll
     public static void init() throws IOException {
         TestClassOutput classOutput = new TestClassOutput();
-        Index index = index(MyService.class, PublicMyService.class, MyItem.class, String.class, CompletionStage.class,
+        Index index = index(MyService.class, PublicMyService.class, BaseService.class, MyItem.class, String.class,
+                CompletionStage.class,
                 List.class);
         ValueResolverGenerator generator = new ValueResolverGenerator(index, classOutput, Collections.emptyMap());
         generator.generate(index.getClassByName(DotName.createSimple(MyService.class.getName())));
@@ -65,8 +65,13 @@ public class SimpleGeneratorTest {
 
     @Test
     public void testWithEngine() throws Exception {
-        Engine engine = Engine.builder().addSectionHelper(new IfSectionHelper.Factory())
-                .addValueResolver(ValueResolvers.thisResolver())
+        try {
+            newResolver("io.quarkus.qute.generator.BaseService_ValueResolver");
+            fail();
+        } catch (ClassNotFoundException | InstantiationException | IllegalAccessException expected) {
+        }
+
+        Engine engine = Engine.builder().addDefaults()
                 .addValueResolver(newResolver("io.quarkus.qute.generator.MyService_ValueResolver"))
                 .addValueResolver(newResolver("io.quarkus.qute.generator.PublicMyService_ValueResolver"))
                 .addValueResolver(newResolver("io.quarkus.qute.generator.MyItem_ValueResolver"))
@@ -75,7 +80,8 @@ public class SimpleGeneratorTest {
                 .build();
         assertEquals(" FOO ", engine.parse("{#if isActive} {name.toUpperCase} {/if}").render(new MyService()));
         assertEquals("OK", engine.parse("{#if this.getList(5).size == 5}OK{/if}").render(new MyService()));
-        assertEquals("Martin NOT_FOUND", engine.parse("{name} {surname}").render(new PublicMyService()));
+        assertEquals("Martin NOT_FOUND OK NOT_FOUND",
+                engine.parse("{name} {surname} {isStatic ?: 'OK'} {base}").render(new PublicMyService()));
         assertEquals("foo NOT_FOUND", engine.parse("{id} {bar}").render(new MyItem()));
     }
 

